src/main/resources/templates/db_scripts/instance-hr-item/write.sql

```diff
@@ -0,0 +1,379 @@
+    RETURN item_metadata;
+  END;
+$$ LANGUAGE plpgsql IMMUTABLE PARALLEL SAFE STRICT;
+
+
+CREATE OR REPLACE FUNCTION set_item_effective_values(holding jsonb, item jsonb)
+  RETURNS jsonb AS $$
+  BEGIN
+    RETURN set_effective_shelving_order(
+          item
+          || jsonb_build_object('effectiveLocationId',
+               COALESCE(item->>'temporaryLocationId', item->>'permanentLocationId', holding->>'effectiveLocationId'))
+          || effective_call_number_components(holding->'new', new.jsonb));
+  END;
+$$ LANGUAGE plpgsql IMMUTABLE PARALLEL SAFE STRICT;
+
+
+CREATE OR REPLACE FUNCTION effective_call_number_components(holding jsonb, item jsonb)
+  RETURNS jsonb AS $$
+  BEGIN
+    RETURN jsonb_build_object('effectiveCallNumberComponents',
+           effective_call_number_component('callNumber', item->>'itemLevelCallNumber', holding->>'callNumber')
+        || effective_call_number_component('prefix', item->>'itemLevelCallNumberPrefix', holding->>'callNumberPrefix')
+        || effective_call_number_component('suffix', item->>'itemLevelCallNumberSuffix', holding->>'callNumberSuffix')
+        || effective_call_number_component('typeId', item->>'itemLevelCallNumberTypeId', holding->>'callNumberTypeId'));
+  END;
+$$ LANGUAGE plpgsql IMMUTABLE PARALLEL SAFE STRICT;
+
+
+CREATE OR REPLACE FUNCTION effective_call_number_component(key text, value1 text, value2 text)
+  RETURNS jsonb AS $$
+  BEGIN
+    IF length(trim(value1)) > 0 THEN
+      RETURN jsonb_build_object(key, value1);
+    END IF;
+    IF length(trim(value2)) > 0 THEN
+      RETURN jsonb_build_object(key, value2);
+    END IF;
+    RETURN jsonb_build_object();
+  END;
+$$ LANGUAGE plpgsql IMMUTABLE PARALLEL SAFE;
+
+
+CREATE OR REPLACE FUNCTION set_effective_shelving_order(item jsonb)
+  RETURNS jsonb AS $$
+  DECLARE
+    call_number text;
+  BEGIN
+    call_number = trim(item->'effectiveCallNumberComponents'->>'callNumber');
+    IF call_number IS NULL OR call_number = '' THEN
+      RETURN item - 'effectiveShelvingOrder';
+    END IF;
+    call_number = concat_ws(' ', call_number,
+        trim2null(item->>'volume'),
+        trim2null(item->>'enumeration'),
+        trim2null(item->>'chronology'),
+        trim2null(item->>'copyNumber'));
+    CASE item->'effectiveCallNumberComponents'->>'typeId'
+      WHEN '03dd64d0-5626-4ecd-8ece-4531e0069f35' THEN call_number = dewey_call_number(call_number);
+      WHEN '95467209-6d7b-468b-94df-0f5d7ad2747d' THEN call_number = lc_nlm_call_number(call_number);  -- LC
+      WHEN '054d460d-d6b9-4469-9e37-7a78a2266655' THEN call_number = lc_nlm_call_number(call_number);  -- NLM
+      WHEN 'fc388041-6cd0-4806-8a74-ebe3b9ab4c6e' THEN call_number = su_doc_call_number(call_number);
+      ELSE NULL;
+    END CASE;
+    RETURN item || jsonb_build_object('effectiveShelvingOrder',
+        concat_ws(' ', call_number, trim2null(item->'effectiveCallNumberComponents'->>'suffix')));
+  END;
+$$ LANGUAGE plpgsql IMMUTABLE PARALLEL SAFE STRICT;
```

Comment:
I am concerned about the prospect of duplicating the logic for creating shelving orders across different locations, specifically in database scripts and with the marc4j library. This approach could potentially lead to inconsistencies in shelving order processing, and maintaining consistency is crucial for ensuring accurate call-number search/browse functionality within the inventory app.

Reply:
It's not described, I can't find any solution for this. Maybe I described my concern not well. We could have inconsistency of shelving order processing. As now we are using marc4j for this processing in both mod-search and mod-inventory-storage after this change moving processing to the database level will require us to support the logic in 2 implementations (db and marc4j).

Reply:
@julianladisch 
Since we have "creating shelving orders logic" implemented in marc4j and in stored functions in the DB, could you please develop tests that compare results produced by marc4j and stored functions for equality?